### DISCUSSION TITLE
Way to go from `cs` to string state name?

### DISCUSSION THREAD

**Discussion by jez** [adrian-thurston/ragel#82] [Ideas]
The lexer I'm building with Ragel has a `trace` mode where it'll print out the tokens it lexes as they're lexed.

One thing I wanted to add to my `lex` function was the ragel state machine's state before and/or after lexing the token. Right now, the state is only manifested as an integer which is hard to understand.

I'd love to have some method to go from `cs` to the state name associated with that state. I'm fine if it's partial (only returns a string if there is an associated named state for that integer).

For example, I have some states in my Ragel machine specification like

```
...

expr_beg := |*
  # ...
*|;

expr_dot := |*
  # ...
*|;

expr_end := |*
  # ...
*|;

...
```

and then in the generated source file that's been compiled from my ragel specification, I see things like:

```
...
static const int lex_en_expr_beg = 664;
static const int lex_en_expr_dot = 463;
static const int lex_en_expr_end = 899;
...
```

I'd love if there was also a generated function I could call like this:

```cpp
static const char * const lex_en_state_name(int cs) {
    switch (cs) {
    // ...
    case lex_en_expr_beg: return "expr_beg";
    case lex_en_expr_dot: return "expr_dot";
    case lex_en_expr_end: return "expr_end";
    // ...
    default: return nullptr;
    }
}
```

so that I could call this in my lexer's `trace` function.

**Comment by jez** [comment]
(Thanks for building Ragel, and for writing such great docs in particular! It's such a rarity these days to have such great, comprehensive docs.)

**Comment by adrian-thurston** [comment]
Hi @jez, yes I have long wanted to do something like this, but never got to it. Note that it would need to return a list of state machine names, as any given state may be derived from multiple names in the ragel input file.

**Comment by burdiyan** [comment]
Oh, this is such a desire of mine! Also ability to display those names instead of numbers in the DOT graph would be very helpful.
